nixos: add option to restrict process information to process owners

joachifm opened the pull request:
This is intended as a conversation starter. This may be either too trivial (the code:boilerplate ratio is ridiculous) or cause too much potential breakage, but I find it to be a nifty option nonetheless, making it more discoverable would be a good thing, IMO.

I have used process information hiding on my desktop for years without problems; I believe it is safe to enable it for typical desktop installations, but have left it disabled by default, as I'm sure it breaks _SOME_ workflow out there (that's the only reason for even making this a module instead of just setting it unconditionally in the stage-2 init).

If the general idea is accepted, I would also add an option to specify a gid that is exempt from the hidepid restriction. This module would then provide a more user-friendly alternative to the severe /proc hardening offered by grsecurity.

Note that this implementation is only a rough draft, I would have to test it properly before it'd be considerd ready.

Thoughts?

copumpkin commented on nixos/modules/security/hidepid.nix:
```diff
@@ -0,0 +1,32 @@
+{
+  options = {
+    security.hideProcessInformation = {
+      enable = mkEnableOption = ''
```
This is a syntax error isn't it? :smile:

joachifm replied:
Indeed it is :) I have yet to actually test this implementation, I'll hold off on doing so until I've tested the waters, to minimize wasted effort.

edolstra commented on nixos/modules/security/hidepid.nix:
```diff
@@ -0,0 +1,32 @@
+  config = mkIf (config.security.hideProcessInformation) {
+    system.activationScripts.hidepid = ''
+      mount -o remount,hidepid=2 /proc
+    '';
```
This won't remount when you *disable* `hideProcessInformation`.

edolstra commented on nixos/modules/security/hidepid.nix:
```diff
@@ -0,0 +1,32 @@
+        this option implies, among other things, that users can can only see
+        their own processes and that information such as command-line arguments
+        is not leaked between users.
+
```
This needs `<para>...</para>`, otherwise you won't get a paragraph break in the manual.

edolstra commented:
:+1: in principle, didn't know this is possible :-)

joachifm commented:
Cool, I'll begin work on a proper implementation

joachifm commented:
I've changed the implementation to a `systemd.service` which undoes the process hiding on stop. Also, I've added an option to make members of a named group exempt from process hiding. Will do some vm testing and then merge (in a squashed form, ofc) barring objections.

abbradar commented:
Maybe we want to do this as an `filesystems."/proc".options` entry instead? `systemd` service seems a wrong tool for the job -- I'm not sure, however.

joachifm commented:
Will that undo the effect on change? I think systemd services are a perfectly good tool for this, why would it not be?

abbradar commented:
Well, what you are effectively doing is adding another mount option for `/proc`. When you want to add an option to the filesystem, usually you don't remount it with a script but add it to `/etc/fstab`, which is represented in NixOS with `filesystems` configuration

joachifm commented:
Okay, it looks like `switch-to-configuration` will indeed remount file systems if necessary so that might work. I will test, however I'm hampered by trunk going through a rebuild right now ... so waiting for stuff to compile.

abbradar commented:
One thing because of which it might _not_ work is that maybe `/proc` is handled differently and you won't get mount options from `/etc/fstab` applied to it on boot.

joachifm commented:
I'm confident that it will pick up things from fstab, that's how I'm doing it on my desktop install

abbradar commented:
`/proc` and maybe `/sys` might be special (other filesystems would definitely work). It depends on when `/proc` is actually mounted -- if it's in stage 1 and is not later remounted, it won't work. I would say this is a bug, though.

joachifm commented:
`systemd-remount-fs.service` should take care of it, but I see that it's disabled in containers, which may be an issue

abbradar commented:
I can check why is it so in containers and if it works without them when I have free time (maybe at the weekend), if you won't beat me to it ~_^.

joachifm commented:
@abbradar I implemented your suggestion here: https://github.com/joachifm/nixpkgs/tree/hidepid-fsoptions. The vm test indicates that this approach won't work in a container.

abbradar commented:
I've tested this. With this patch, it works in a container _but_ doesn't switch configuration properly. This is because we need to fix `switch-to-configuration.pl` so that it would restart `systemd-remount-fs.service` when `/proc` or `/sys` are touched (this looks like a bug).

```
diff --git a/nixos/modules/security/hidepid.nix b/nixos/modules/security/hidepid.nix
index 0e8bf1c..1795878 100644
--- a/nixos/modules/security/hidepid.nix
+++ b/nixos/modules/security/hidepid.nix
@@ -33,9 +33,13 @@ in
     # Note: these options must be set regardless of whether the module is
     # "enabled" to ensure that the effect of enabling process information hiding
     # is undone when the module is disabled.
-    fileSystems."/proc".options = [
-      ''hidepid=${if cfg.enable then "2" else "0"}''
-      ''gid=${if (cfg.enable && cfg.procGID != null) then "${toString cfg.procGID}" else "0"}''
-    ];
+    fileSystems."/proc" = {
+      device = "proc";
+      fsType = "proc";
+      options = [
+        ''hidepid=${if cfg.enable then "2" else "0"}''
+        ''gid=${if (cfg.enable && cfg.procGID != null) then "${toString cfg.procGID}" else "0"}''
+      ];
+    };
   };
 }
diff --git a/nixos/modules/system/boot/systemd.nix b/nixos/modules/system/boot/systemd.nix
index a3c8352..4505ad7 100644
--- a/nixos/modules/system/boot/systemd.nix
+++ b/nixos/modules/system/boot/systemd.nix
@@ -783,7 +783,6 @@ in
     systemd.services.systemd-binfmt.wants = [ "proc-sys-fs-binfmt_misc.automount" ];
 
     # Don't bother with certain units in containers.
-    systemd.services.systemd-remount-fs.unitConfig.ConditionVirtualization = "!container";
     systemd.services.systemd-random-seed.unitConfig.ConditionVirtualization = "!container";
 
   };
(END)
```

abbradar commented:
To clarify: I think that enabling `systemd-remount-fs` in containers should be harmless and it was disabled just to avoid previously useless services. I might be wrong though, @edolstra may know more.

joachifm commented:
I agree, it makes sense to have `systemd-remount-fs` enabled, it exists only to remount root + API file systems so that fstab options are picked up.

abbradar commented:
Going to sleep now; I'm now Perl-proficient at all but can look at fixing this in our switch script later.

cstrahan commented on nixos/modules/security/hidepid.nix:
```diff
@@ -0,0 +1,54 @@
+          By default, all users except root are subject to process information
+          hiding. Services that require access to process information should be
+          executing with this GID added to their supplementary groups (or as the
+          primary group). This is achived by setting
```
s/achive/achieved/

cstrahan commented:
This certainly seems useful to me :+1:

offlinehacker commented:
Nice :)

abbradar commented:
I've added proper handling of `/proc` and `/sys` to `switch-to-configuration.pl` in https://github.com/joachifm/nixpkgs/pull/1. `fileSystems` approach works now on ordinary configurations and in containers but it doesn't work in VMs because of (some overrides)[https://github.com/NixOS/nixpkgs/blob/master/nixos/modules/virtualisation/qemu-vm.nix#L420]. Maybe we want to go with @joachifm's original idea since it works everywhere including tests (and can thus be tested)...

abbradar commented on nixos/modules/security/hidepid.nix:
```diff
@@ -33,6 +33,12 @@ in
+         message = "the group named at security.hideProcessInformation.procGroup does not exist!";
+       }
+     ];
+
```
This would not necessarily be so -- `mutableUsers` is on by default so users can add their own groups circumventing NixOS configuration.

joachifm replied:
What is the correct way of approaching it?

abbradar replied:
Sorry, I didn't understand at this point that you need GID (and need to resolve it somehow from the group name).

joachifm replied:
I have considered changing the option to simply indicating whether a group named "proc" should be created and be made exempt instead of letting the user name a group of their choosing. It's actually what I'd prefer; I felt that the current approach has smaller impact on the rest of the system, however.

abbradar replied:
I actually don't think it would be an issue. We already allocate separate GIDs for all services (some even have two and maybe more!) -- added group dedicated to this shouldn't be a problem and would cut much of the complexity.

joachifm replied:
It would also have the benefit of working even if group info was not available for some reason (I *think* that'll never be the case, but still). I will switch to a static group id, then.

abbradar commented on nixos/modules/security/hidepid.nix:
```diff
@@ -36,7 +36,12 @@ in
+      # but we cannot be sure that systemd-remount-fs will run; ordering this
+      # service after local-fs.target will work in either case
+      # (systemd-remount-fs is ordered before local-fs, by default).
       after = [ "local-fs.target" ];
```
Wny not add `systemd-remount-fs.service` to `after` then? It would order them properly if `systemd-remount-fs` runs without requiring it to run it all.

abbradar commented on nixos/modules/security/hidepid.nix:
```diff
@@ -17,10 +17,9 @@ in
-        example = 1337;
+      procGroup = mkOption {
+        type = types.nullOr types.str;
+        example = "proc";
```
Oh, I see now -- we need to have GID in mount options. What about allocating a dedicated group in `ids.nix` and using it instead of `procGroup`?

joachifm committed:
nixos: add optional process information hiding

This module adds an option `security.hideProcessInformation` that, when
enabled, restricts access to process information such as command-line
arguments to the process owner.  The module adds a static group "proc"
whose members are exempt from process information hiding.

Ideally, this feature would be implemented by simply adding the
appropriate mount options to `fileSystems."/proc".fsOptions`, but this
was found to not work in vmtests. To ensure that process information
hiding is enforced, we use a systemd service unit that remounts `/proc`
after `systemd-remount-fs.service` has completed.

To verify the correctness of the feature, simple tests were added to
nixos/tests/misc: the test ensures that unprivileged users cannot see
process information owned by another user, while members of "proc" CAN.

Thanks to @abbradar for feedback and suggestions.
